packages/angular/ssr/src/routes/route-config.ts

```diff
@@ -223,3 +262,78 @@ export function provideServerRoutesConfig(
     },
   ]);
 }
+
+/**
+ * Sets up the necessary providers for configuring server routes.
+ * This function accepts an array of server routes and optional configuration
+ * options, returning an `EnvironmentProviders` object that encapsulates
+ * the server routes and configuration settings.
+ *
+ * @param routes - An array of server routes to be provided.
+ * @param features - (Optional) server routes features.
+ * @returns An `EnvironmentProviders` instance with the server routes configuration.
+ *
+ * @see {@link ServerRoute}
+ * @see {@link withAppShell}
+ * @developerPreview
+ */
+export function provideServerRouting(
+  routes: ServerRoute[],
+  ...features: ServerRoutesFeature<ServerRoutesFeatureKind>[]
+): EnvironmentProviders {
+  const config: ServerRoutesConfig = { routes };
+  const hasAppShell = features.some((f) => f.ɵkind === ServerRoutesFeatureKind.AppShell);
```

Comment:
Question: Do we need to check the kind here or could we just look for a route containing the app shell route?

Related question: Does framework typically implement this `withFeature` pattern using an object with `kind` to disambiguate the relevant feature? I always assumed it just included a bunch of providers and let the runtime optionally inject and call those providers if they are present. Basically, I'm wondering if there's a path to make `withAppShell` just return a list of providers rather than needing a special `kind` enum.

Reply:
> Do we need to check the kind here or could we just look for a route containing the app shell route?

The route config will not contain the app-shell route, the route will be added during runtime.

> Does framework typically implement this withFeature pattern using an object with kind to disambiguate the relevant feature

Yes, this is the pattern used in the framework https://github.com/angular/angular/blob/fd7ee47bf702029e481ad40362684f0a01ef2ec0/packages/router/src/provide_router.ts#L110-L123, https://github.com/angular/angular/blob/main/packages/common/http/src/provider.ts and all others. I think the goal here, is to actually make sure users to not add these providers directly in their config example `providers: [withFeature()]`.

Reply:
Interesting, I get we don't want users to write `providers: [withFeature]`. I was envisioning something like:

```typescript
interface Feature {
  providers: Provider[];
}
```

And then we'd just append the providers without worrying about a `kind`.

I'll defer to you whether the `kind` is worth keeping, seems like it was useful for the app router at least.

Reply:
The `kind` is helpful for filtering providers and performing additional actions based on its value.